Getting Started with Contributing to RDKit (#7813)
rdkit/rdkit: Docs/Book/GettingStartedWithContributing.md

- The [RDKit Book](https://www.rdkit.org/docs/RDKit_Book.html) gives more technical details about how various parts of the RDKit work.  
- Not strictly speaking documentation but we'll include it here for completeness, Greg writes a [Blog](https://greglandrum.github.io/rdkit-blog/) highlighting in more detail how to use new and existing bits of the RDKit.

### How to Write RDKit Documentation

The RDKit documentation is written mostly in [reStructuredText](https://docutils.sourceforge.io/rst.html). Full tutorials can be found online, but a brief overview will be given here to get you started. 

*Headings*

  ```
  This is a TITLE
  %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

  This is a HEADING 
  **************

  This is a SUBHEADING
  ========================
  ```

*Code Examples*

The RDKit docs support inclusion of Sphinx doctests to allow the result of a code example to appear in the text. 

  ```
  .. doctest::

    >>> mol = Chem.MolFromSmiles('CO(C)C')
    >>> print(print(mol.GetNumAtoms()))
    4
  ```

*Hyperlinks*

  ```
  `RDKit Website`_

  .. _RDKit Website: https://www.rdkit.org/
  ```

*Images*

  ```
  .. image:: pathtoimage/name.png
  ```

> ### Tutorials

Reviewer: The tutorials repo seems moribund - no updates since last year - so it's probably better to just drop this section.

Author: That's a shame - I think mention of them somewhere is still useful for a "beginners" guide, will add a link beside the description of the cookbook above.

- tutorials are written in [Jupyter](https://jupyter.org/) format
- naming of tutorials follows the convention `XXX_NAME.ipynb` where XXX is a zero-padded number and NAME gives a brief idea of what is contained in the tutorial
- include a cell with `@TAGS:` and the relevant tags appropriate to the tutorial
- content can include brief explanations with links to more detail to complement the code examples

Examples of how to use RDKit in C++ are found [here](https://github.com/rdkit/rdkit/tree/master/Docs/Book/C%2B%2BExamples
). 

## Contributing to the Code - Python 

### Building RDKit for Development

If you also have, or plan to, contribute C++ code to the RDKit, the same set-up can be used for Python contributions (see instructions below). However if you intend only to contribute pure Python, which may be easiest for beginners, [Greg recently created some easier to follow recipes](https://greglandrum.github.io/rdkit-blog/posts/2020-03-30-setting-up-an-environment.html) to simplify the start-up process. To build the environment you will need:

> - Local installation of git and either [Anaconda](https://www.anaconda.com/download) or [Miniconda](https://docs.anaconda.com/miniconda/miniconda-install/)

Reviewer: No one should ever install the full anaconda distro, so I suggest changing this to:
```suggestion
- Local installation of git and either [miniconda](https://docs.anaconda.com/miniconda/miniconda-install/) or [miniforge](https://github.com/conda-forge/miniforge)
```

- Clone your fork of the RDKit GitHub repository following the instructions given above
- Change to the directory containing the local clone and set the RDBASE environment variable:
    ```
    cd rdkit
    export RDBASE=`pwd`
    ```
- Create a conda environment containing the most recent RDKit version (you can give this whatever name you like and adjust Python version as needed):
    ```
    conda create -y -n py311_rdkit_beta python=3.11
    conda activate py311_rdkit_beta

>     conda install -y -c conda-forge rdkit pytest

Reviewer: conda create -y -n py312_rdkit_beta python=3.12
conda activate py312_rdkit_beta
conda install -y -c conda-forge rdkit pytest

    ```
- Copy the RDKit binary components from that environment into our local clone of the RDKit repo (remembering to adjust the path if you have changed the Python version):
  - **Mac and Linux**
      ```

>       cd $CONDA_PREFIX/lib/python3.11/site-packages/rdkit

Reviewer: ```suggestion
      cd $CONDA_PREFIX/lib/python3.12/site-packages/rdkit
```

      rsync -a -m --include '*/' --include='*.so' --include='inchi.py' --exclude='*' . $RDBASE/rdkit
      ```
  - **Windows**
      ```

>       cd $CONDA_PREFIX/lib/python3.11/site-packages/rdkit

Reviewer: ```suggestion
      cd $CONDA_PREFIX/lib/python3.12/site-packages/rdkit
```

      find . -name '*.pyd' -exec cp --parents \{\} $RDBASE/rdkit \; 
      cp Chem/inchi.py $RDBASE/rdkit/Chem
      ```
- Set the PYTHONPATH and test everything is working my importing RDKit:
    ```
    export PYTHONPATH="$RDBASE"
    cd $RDBASE/rdkit
    python -c 'from rdkit import Chem;print(Chem.__file__)'
    ```
- You can also test this by running the Python tests:
    ```

>     cd $RDBASE/rdkit/Chem
>     python -m pytest

Reviewer: ```suggestion
    cd $RDBASE/rdkit
    python -m pytest
```

    ```

Any issues encountered with these instructions can be reported in [this thread](https://github.com/rdkit/rdkit/issues/3052).

### Where Should I Put My Code? 

The RDKit Python code lives in the `rdkit/rdkit` directory. Most contributions (and their corresponding unit tests) will be added to `Chem` and the relevant sub-directory, but check with the maintainers if you are unsure. Files containing any examples used in tests should be placed in a sub-directory called `test_data`.

### Coding Standards

There is currently no official style guide or linter setup that you should use for Python. However, there are a few rules to adhere to so that the code and API design stays relatively coherent (sidenote: it is currently not):
- Use `CamelCase` for `PackageNames`, `ModuleNames`, `ClassNames`, `MethodNames` and `FunctionNames`, but note the different style for `functionAttributeNames` and `methodAttributeNames`
 - These recommendations are made based on the most commonly found patterns in the `Chem` module. However, you may see small inconsistencies here and then...
   <img src="../Book/images/style.png" width="500">

> - Use `snake_case` names for local variables inside methods and functions. However, `justlowercase` is also fine.

Reviewer: please remove this one

- Use multi-line Python strings to document modules and methods, i.e.:
    ```python
    def xy():
      """ This is a documentation string.
      
      It is continued here...
      """
    ```

> - Use double quotes (`"String Example"`) for string literals.

Reviewer: please remove this one

Full installation instructions for the RDKit can be found [here](https://github.com/rdkit/rdkit/blob/master/Docs/Book/Install.md). Development in C++ requires a build from source. A summary of the required steps is as follows:
- Clone the RDKit GitHub repository following the instructions above
- Set the required paths:
    ```
    export RDBASE=$(pwd) # from your local rdkit directory
    export DYLD_LIBRARY_PATH=$RDBASE/build/lib
    export CPATH= /USER/RESTOFPATHTO/rdkit/Code
    export LIBRARY_PATH=/USER/RESTOFPATHTO/rdkit/lib
    export PYTHONPATH=$RDBASE
    export PYROOT=/USER/RESTOFPATHTO/miniconda3
    ```
- create and move to the build directory:
    ```
    mkdir build
    cd build
    ```
- run cmake with the relevant flags e.g.
  ```
  cmake -DRDK_BUILD_PYTHON_WRAPPERS=OFF ..
  ```
  if you don't want to build the Python wrappers. The full list of options can be found in the [CMakeLists.txt](https://github.com/rdkit/rdkit/blob/master/CMakeLists.txt) file.
- run make:
  ```
  make install -j10
  ```
- run the tests to check the build has been successful:
  ```
  RDBASE=$RDBASE DYLD_FALLBACK_LIBRARY_PATH="$RDBASE/lib:$PYROOT/lib" PYTHONPATH=$RDBASE ctest -j 10
  ```

> Note: on MacOS for v14.0 onwards some of the changes caused problems with the standard CMake spell. This [discussion thread](https://github.com/rdkit/rdkit/discussions/6859) provides some useful solutions in this case.

Reviewer: this info should be in the Install.md script, so please remove it here

### Where Should I Put My Code? 

The RDKit C++ code lives in `rdkit/Code`, with the majority of the functionality sitting in the `GraphMol` sub-directory. If the code being contributed interfaces with another project it should go in `rdkit/External` and in an appropriate sub-directory. Again if you are unsure where your submission should go check with the maintainers!

### Coding Standards

As discussed in the Python section above, there is no official formal style guide for RDKit, but we recommend following the suggestions below:
- Generally the toolkit uses camelCase. Class and namespace names start with an upper-case letter (e.g. ExplicitBitVect), function and method names starting with a lower-case letter (e.g. getNumAtoms()).

> - Local variable names are usually lower-case, but sometimes use either snake_case or camelCase.

Reviewer: please remove this one

> - We follow the formatting suggestions from [Google's C++ style guidelines](https://google.github.io/styleguide/cppguide.html).

Reviewer: please remove this one

- Use of an auto-formatter (e.g. [clang-format](http://clang.llvm.org/docs/ClangFormat.html)) is strongly recommended. This can be integrated with editors, but note that at least version 3.8 of clang-format is required as the configuration files are not backwards compatible. There is a .clang-format config file at the root of the repository; code layout should be based on this configuration.

> - The public classes / functions exposed in the API are typically defined in the header file, and the underlying method in the C++ file of the same name.

Reviewer: ```suggestion
- The public classes / functions exposed in the API are typically declared in the header file, and the underlying definition is in the C++ file of the same name.
```

- Use multi-line comments in the header file to document code:
    ```c++
    /*!
    this is a documentation string
    */
    ```

We would also recommend checking your code for memory leaks with [Valgrind](https://valgrind.org/) before opening a pull request.

### Tests

> The RDKit C++ Tests use the [Catch2 framework](https://github.com/catchorg/Catch2). The tests generally are placed in a file named "catch_tests.cpp" in the same directory as the corresponding code. Small changes and bug fixes may only need the addition of a new TEST\_CASE to an existing test set rather than a new file.

Reviewer: ```suggestion
The RDKit C++ Tests use the [Catch2 framework](https://github.com/catchorg/Catch2). The tests generally are placed in a file named "catch_tests.cpp" in the same directory as the corresponding code. Unless you are adding a significant new feature, it's generally best to add the new TEST\_CASE to an existing test set rather than creating an entirely new file.
```

For example, to test the atom count function:

```c++
#include "RDGeneral/test.h"
#include <catch2 /catch_all.hpp>
#include <GraphMol /SmilesParse/SmilesParse.h>

using namespace RDKit;

TEST_CASE("test basic Mol features", "[ROMol]") {
  SECTION("basics") {
    {
    // shortcut to create a molecule from SMILES
    auto m = "CCO"_smiles;
    // check that the molecule was created successfully
    REQUIRE(m);
    // test some functionality
    int nAtoms = m->getNumAtoms();
    CHECK(nAtoms == 3);
    }
  }
}
```

> When adding new tests the CMakeLists.txt file of the respective folder will also need updated to call the test and link the required libraries (e.g. the above example needs the FileParsers library):

Reviewer: ```suggestion
When adding new test file the CMakeLists.txt file of the respective folder will also need updated to call the test and link the required libraries (e.g. the above example needs the FileParsers library):
```

```
rdkit_catch_test(dummyTest catch_dummy.cpp LINK_LIBRARIES FileParsers)
```

### Documentation

Documentation for the [C++ API](https://www.rdkit.org/docs/cppapi/) is generated from the docstrings, which should be placed above the corresponding function. We'd recommend these take the following format:

```c++
/*!
    A brief description of what the function does

    Reference / Source (if there is one)

    \param name: what the parameter is
    \param name2: what the parameter is
    etc...
  */
```

> Where relevant, example useage can also be added to the [Getting Started with C++ Guide](https://www.rdkit.org/docs/GettingStartedInC%2B%2B.html) as part of the pull request.

Reviewer: please remove this one

Author: removed the documentation section and example from the tests, but have left a link to the catch2 framework and where to add new tests (I didn't know this when doing the DCLV stuff!)
